### PR TITLE
Update router/kfctlClient/kfctlServer API

### PR DESCRIPTION
**Pull request by kunmingg**
* router take user input -> build kfdef -> send to kfctlServer
* update kfdef interface so object fields are editable. 

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/4399)

### CODE REVIEW

File: bootstrap/cmd/bootstrap/app/kfctlClient.go [hunk 1]
```diff
@@ -159,27 +167,29 @@ func (c *KfctlClient) CreateDeployment(ctx context.Context, req kfdefs.KfDef) (*
...
-	if ok {
-		return nil, resErr
-	}
-
-	log.Info("Response is not type *httpError")
-
-	pRes, _ := Pformat(resp)
-	log.Errorf("Received unexpected response; %v", pRes)
-	return nil, fmt.Errorf("Received unexpected response; %v", pRes)
+func (c *KfctlClient) GetLatestStatus(req KfctlRequest) (*kfconfig.Status, error) {
```
Comment (jlewi) [inline bootstrap/cmd/bootstrap/app/kfctlClient.go]: Return type should full KfConfig not just status
Add a TODO about why its commented out.

File: bootstrap/cmd/bootstrap/app/router.go [hunk 2]
```diff
@@ -40,6 +39,47 @@ type kfctlRouter struct {
...
+}
+
+// For GCP cloud endpoint
+type EndpointConfig struct {
+	BasicAuth BasicAuth `json:"basicAuth,omitempty"`
+	IAP       IAP       `json:"iap,omitempty"`
+}
+
+// KfctlRequest is the request body for CreateDeployment and GetLatestStatus API
+type KfctlRequest struct {
```
Comment (jlewi) [inline bootstrap/cmd/bootstrap/app/router.go]: Lets call this something like ClickToDeployRequest
To indicate that its an internal request.  And lets expand the comment to explain that this is an "internal" API only for click to deploy front end.

File: bootstrap/cmd/bootstrap/app/kfctlClient.go [hunk 3]
```diff
@@ -21,6 +22,13 @@ type KfctlClient struct {
 	getEndpoint    endpoint.Endpoint
 }
 
+type KfctlService interface {
```
Comment (jlewi) [inline bootstrap/cmd/bootstrap/app/kfctlClient.go]: Please add a TODO per discussion that this should be replaced by a versioned API

KfctlServiceV1Alpha1
KfctlServiceV1Beta1
KfctlServiceV1

That would use the corresponding versioned  KFDef spec as arguments.

File: bootstrap/pkg/kfupgrade/kfupgrade.go [hunk 4]
```diff
@@ -242,7 +242,7 @@ func MergeKfCfg(oldKfCfg *kfconfig.KfConfig, newKfCfg *kfconfig.KfConfig) {
 }
 
 func (upgrader *KfUpgrader) Generate() error {
-	kfApp, err := coordinator.NewLoadKfAppFromURI(upgrader.TargetPath)
+	kfApp, err := coordinator.LoadKfAppCfgFile(upgrader.TargetPath)
```
Comment (jlewi) [inline bootstrap/pkg/kfupgrade/kfupgrade.go]: Why are you changing kfupgrade? Do you just need to rebase?

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 5]
```diff
@@ -2160,11 +2160,6 @@ func (gcp *Gcp) Generate(resources kftypesv3.ResourceEnum) error {
 		}
 	}
 
-	// TODO(jlewi): Why is this hard coded here?
```
Comment (jlewi) [inline bootstrap/pkg/kfapp/gcp/gcp.go]: Why is this code changing?

Comment (kunmingg) [reply bootstrap/pkg/kfapp/gcp/gcp.go]: This param is not being used anymore

File: bootstrap/cmd/bootstrap/app/router.go [hunk 6]
```diff
@@ -40,6 +40,47 @@ type kfctlRouter struct {
...
+}
+
+// For GCP cloud endpoint
+type EndpointConfig struct {
+	BasicAuth BasicAuth `json:"basicAuth,omitempty"`
+	IAP       IAP       `json:"iap,omitempty"`
+}
+
+// C2DRequest is the request body from click to deploy frontend
+type C2DRequest struct {
```
Comment (kunmingg) [inline bootstrap/cmd/bootstrap/app/router.go]: request body from frontend for v0.7

File: bootstrap/cmd/bootstrap/app/router.go [hunk 7]
```diff
@@ -40,6 +40,47 @@ type kfctlRouter struct {
...
+	Zone string
+
+	// kubeflow Version
+	Version string
+
+	shareAnonymousUsage bool
+
+	// Temporary service account access Token
+	Token string
+}
```
Comment (kunmingg) [inline bootstrap/cmd/bootstrap/app/router.go]: request body from frontend for v0.7

File: bootstrap/cmd/bootstrap/app/kfctlClient.go [hunk 8]
```diff
@@ -75,9 +79,45 @@ func NewKfctlClient(instance string) (KfctlService, error) {
...
+	kfconfigIns.Spec.ConfigFileName = filepath.Base(req.ConfigFile)
+
+	// Fill in user input
+	kfconfigIns.Name = req.Name
+	// Not passing a pointer interface is a common cause of deserialization problems
+	pluginSpec := &gcp.GcpPluginSpec{}
+
+	err = kfconfigIns.GetPluginSpec(gcp.GcpPluginName, pluginSpec)
+	if err != nil {
+		deployReqCounter.WithLabelValues("INTERNAL").Inc()
```
Comment (jlewi) [inline bootstrap/cmd/bootstrap/app/kfctlClient.go]: What does INTERNAL mean here?

File: bootstrap/cmd/bootstrap/app/kfctlClient.go [hunk 9]
```diff
@@ -152,34 +192,36 @@ func (c *KfctlClient) CreateDeployment(ctx context.Context, req kfdefs.KfDef) (*
...
-	if ok {
-		return nil, resErr
-	}
-
-	log.Info("Response is not type *httpError")
-
-	pRes, _ := Pformat(resp)
-	log.Errorf("Received unexpected response; %v", pRes)
-	return nil, fmt.Errorf("Received unexpected response; %v", pRes)
+func (c *KfctlClient) GetLatestKfdef(req kfconfig.KfConfig) (*kfconfig.KfConfig, error) {
```
Comment (jlewi) [inline bootstrap/cmd/bootstrap/app/kfctlClient.go]: Please add a TODO explaining why this commented out.

File: bootstrap/cmd/bootstrap/app/router.go [hunk 10]
```diff
@@ -439,18 +468,19 @@ func (r *kfctlRouter) CreateDeployment(ctx context.Context, req kfdefs.KfDef) (*
...
-	}
-	address := fmt.Sprintf("http://%v.%v.svc.cluster.local:80", name, r.namespace)
-	log.Infof("Creating client for %v", address)
-	c, err := NewKfctlClient(address)
-	return c.GetLatestKfdef(req)
+	// TODO
+	return &kfdefs.KfDef{}, nil
 }
+//
+//// GetLatestKfdef returns latest kfdef status.
```
Comment (jlewi) [inline bootstrap/cmd/bootstrap/app/router.go]: Please add a TODO explaining why this is commented out.

The best thing would be to link to an issue.

File: bootstrap/pkg/apis/apps/configconverters/converters.go [hunk 11]
```diff
@@ -26,21 +24,9 @@ const (
...
-
-// LoadConfigFromURI reads the kfdef from a remote URI or local file,
-// and returns the kfconfig.
-// It will set the AppDir and ConfigFilename in kfconfig:
-//   AppDir = cwd if configFile is remote, or it will be the dir of configFile.
-//   ConfigFilename = the file name of configFile.
-func LoadConfigFromURI(configFile string) (*kfconfig.KfConfig, error) {
+// This function should be thread safe and should not write to disk.
+// LoadFileFromURI: get file content from URI (lcoal or remote) and return as []byte
+func LoadFileFromURI(configFile string) ([]byte, error) {
```
Comment (jlewi) [inline bootstrap/pkg/apis/apps/configconverters/converters.go]: Why are we returning things as []byte? If we want to return a variable type couldn't we do interface{} and then the caller can do a type assertion?

File: bootstrap/pkg/apis/apps/configconverters/converters.go [hunk 12]
```diff
@@ -50,61 +36,27 @@ func LoadConfigFromURI(configFile string) (*kfconfig.KfConfig, error) {
...
-		}
+		return ioutil.ReadAll(resp.Body)
+	} else {
+		// If config is lcoal, read file.
+		return ioutil.ReadFile(configFile)
 	}
+}
 
+// KfdefByteToKfConfig takes Kfdef in bytes and convert to KfConfig
+func KfdefByteToKfConfig(KfdefByte []byte, appDir string) (*kfconfig.KfConfig, error) {
```
Comment (jlewi) [inline bootstrap/pkg/apis/apps/configconverters/converters.go]: Per comment above do we need to do this if we use a type interface{} ? Then we could just to type assertion?

**Comment by jlewi** [comment]
@kunmingg High level comment; we have two competing goals here.

1. We want to minimize changes to logic shared with kfctl (eg. configconverters) to avoid destabilizing the release
1. We want to get something in quick to fix C2D

Is there a way to avoid making changes to the code shared with kfctl? i.e. can we punt any future cleanup to future PRs which won't be rushed?

IIUC the main need to change configconverters is because it only supports loading a KFDef from a URI but in KFCTLServer the KFDef comes from the request.

Instead of modifying configconverters in this PR could KFCTLServer just do the following

1. Create APPDIR
2. Write the request to ${APPDIR}/${CONFIGNAME}
3. Call one of the existing kfctl methods using ${APPDIR}/${CONFIGNAME}?

File: bootstrap/pkg/apis/apps/configconverters/converters.go (hunk 11 above)
Comment (kunmingg) [reply bootstrap/pkg/apis/apps/configconverters/converters.go]: The converter interface now takes file content in []byte as input:
https://github.com/kubeflow/kubeflow/blob/master/bootstrap/pkg/apis/apps/configconverters/converters.go#L22
If use interface{} then need to convert it back later

**Comment by kunmingg** [comment]
@jlewi 
Agree, some configconverters change are unnecessary, reverting them locally.

**Comment by jlewi** [comment]
@kunmingg is this ready for another look? If it is could you remove WIP form the title?

**Comment by kunmingg** [comment]
@jlewi 
Ready for another look, 
going to partially revert changes in `configconverters`

**Comment by jlewi** [comment]
@kunmingg Why only partial revert of configconverters? What are the needed changes to configconverters?

**Comment by kunmingg** [comment]
@jlewi 
Now for `configconverters` the only change is:
1. http get remote file instead of download to disk
2. allow specify app-dir when build kfconfig

**Comment by jlewi** [comment]
@kunmingg see comment on the configconverters file.

**Comment by kunmingg** [comment]
`req.ConfigFile = https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_gcp_basic_auth.0.7.0.yaml`
Here we need converter to read config template and build kfconfig.

**Comment by jlewi** [comment]
@kunmingg nit; would you mind updating the PR description? I think we are no longer building KFConfig but instead building KFDefs?

**Comment by jlewi** [comment]
Thanks @kunmingg it looks like you changed coordinator.go e.g (KfConfigGetter) to return a KFConfig. Can we avoid exposing KFConfig? 

Could we have Coordinator expose versioned Getters/Setters e.g.

KFDefv1beta1Getter
KFDefv1Getter?

File: bootstrap/cmd/bootstrap/app/kfctlClient.go (hunk 8 above)
Comment (kunmingg) [reply bootstrap/cmd/bootstrap/app/kfctlClient.go]: INTERNAL mark the error as true error, instead of false positives like invalid input fields.

**Comment by jlewi** [comment]
@kunmingg per @richardsliu email looks like we just cut over to kubeflow/kfctl.
So unfortunately you'll need to patch changes into that repo.

There's a recipe [here](https://github.com/kubeflow/kfctl/issues/7#issuecomment-531904794) for patching changes over.

**Comment by jlewi** [comment]
@kunmingg overall the changes (i.e. creating KFGetter that just returns a KFDefV1beta1) looks really nice.

I think the main comments are

1. Open a new PR in kubeflow/kfctl
1. Only include the changes in pkg/ in the first PR

Sorry for all the hassle.

**Comment by kunmingg** [comment]
Close as it was split into 2 PRs which is merged.

**Pull request closed without merging**